### PR TITLE
Move ossl_asn1_string_to_time_t() to libtestutil

### LINKED ISSUE #22965: Move ossl_asn1_string_to_time_t() to libtestutil?
Not a bug per se, but a trivial observation... Would it not make more sense to locate the `ossl_asn1_string_to_time_t()` function in `libtestutil` instead of `libcrypto`? It is currently only used by tests and not by anything in libcrypto, Tests do not necessarily have access to this function normally as it is an internal function of libcrypto (would avoid having to forcibly link the involved tested with the static libcrypto).

### PR DESCRIPTION
**Pull request by coolshrid**
Fixes #22965
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/testutil/helper.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "../testutil.h"
+
+time_t test_asn1_string_to_time_t(const char *asn1_string)
+{
+    ASN1_TIME *timestamp_asn1 = NULL;
+    struct tm *timestamp_tm = NULL;
+#if defined(__DJGPP__)
+    char *tz = NULL;
+#elif !defined(USE_TIMEGM)
```
Comment (InfoHunter) [inline test/testutil/helper.c:21]: This `USE_TIMEGM` thing is defined in the original `a_time.c`, so you need either move that definition lines here, or just `defined(__FreeBSD__) || defined(__wasi__)`

File: test/testutil/helper.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    time_t timestamp_local;
+#endif
+    time_t timestamp_utc;
+
+    timestamp_asn1 = ASN1_TIME_new();
```
Comment (InfoHunter) [inline test/testutil/helper.c:26]: Need to check the return value of `ASN1_TIME_new()`?

**Comment by t8m** [comment]
CI failures are relevant

File: test/testutil/helper.c (hunk 1 above)
Comment (t8m) [reply test/testutil/helper.c:21]: Also the timezone definition for Windows must be moved.

**Comment by coolshrid** [comment]
Fixed build issues with Windows env.

File: test/testutil/helper.c [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    if(timestamp_asn1 == NULL)
```
Comment (t8m) [inline test/testutil/helper.c:38]: Please add a space after `if`

**Review by t8m — CHANGES_REQUESTED** [review]
Just one nit to be fixed

File: test/testutil/helper.c (hunk 3 above)
Comment (coolshrid) [reply test/testutil/helper.c]: Fixed formatting issue.

**Comment by coolshrid** [comment]
Fixed code formatting issues.

**Comment by t8m** [comment]
Can you please drop the merge commit? `git reset --hard HEAD~ && git push --force` should help.

**Comment by coolshrid** [comment]
Done

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Unfortunately there is a merge commit again. Could you please drop it?

**Comment by coolshrid** [comment]
Sure, can you help me understand what needs to be dropped?

Thanks

From: Tomáš Mráz ***@***.***>
Date: Wednesday, January 24, 2024 at 03:08
To: openssl/openssl ***@***.***>
Cc: shridhar ***@***.***>, Author ***@***.***>
Subject: Re: [openssl/openssl] Move ossl_asn1_string_to_time_t() to libtestutil (PR #23269)


Unfortunately there is a merge commit again. Could you please drop it?

**Comment by coolshrid** [comment]
Do I have to run the following command again?

> git reset --hard HEAD~ && git push --force

**Comment by t8m** [comment]
> Do I have to run the following command again?
> 
> > git reset --hard HEAD~ && git push --force

Yes, please. I do not know how these merge commits get added on your PR's perhaps it is some kind of auto-sync happening?

**Comment by coolshrid** [comment]
Done. Ran the following

> git reset --hard HEAD~ && git push --force

File: test/testutil.h [hunk 4]
```diff
@@ -649,4 +649,10 @@ X509 *load_cert_der(const unsigned char *bytes, int len);
 STACK_OF(X509) *load_certs_pem(const char *file);
 X509_REQ *load_csr_der(const char *file, OSSL_LIB_CTX *libctx);
 
+/*
+  moved from ./include/crypto/asn1.h  
+  time_t ossl_asn1_string_to_time_t(const char *asn1_string);
+*/
```
Comment (hlandau) [inline test/testutil.h:655]: I don't think this comment is really necessary.

**Review by hlandau — APPROVED** [review]

File: test/testutil.h (hunk 4 above)
Comment (t8m) [reply test/testutil.h:655]: Yeah, it is also wrongly formatted. Please just drop it.

Comment (tom-cosgrove-arm) [reply test/testutil.h:655]: If this block (which has trailing whitespace on line 653) is deleted, I can approve this PR

File: test/testutil.h [hunk 5]
```diff
@@ -648,5 +648,5 @@ X509 *load_cert_pem(const char *file, OSSL_LIB_CTX *libctx);
 X509 *load_cert_der(const unsigned char *bytes, int len);
 STACK_OF(X509) *load_certs_pem(const char *file);
 X509_REQ *load_csr_der(const char *file, OSSL_LIB_CTX *libctx);
-
-#endif                          /* OSSL_TESTUTIL_H */
+time_t test_asn1_string_to_time_t(const char *asn1_string);
+#endif                          /* OSSL_TESTUTIL_H */
```
Comment (tom-cosgrove-arm) [inline test/testutil.h:652]: Thank you for removing the comment block (such commentary about changes in the code base happens naturally by looking at the git commit history).

However, this last change has lost the trailing newline (\n) in the file - can you put it back, please? The CI is complaining:

```
In file included from test/testutil/apps_shims.c:12:
test/testutil/../testutil.h:652:54: error: no newline at end of file [-Werror,-Wnewline-eof]
#endif                          /* OSSL_TESTUTIL_H */
                                                     ^
1 error generated.
```

Comment (coolshrid) [reply test/testutil.h:652]: Updated.

Comment (tom-cosgrove-arm) [reply test/testutil.h:652]: Err, have you pushed the change? It's not visible on GitHub here

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
@t8m can you reconfirm?

**Comment by coolshrid** [comment]
Yes (pushed with commit #1d7168a70c

[image.png]

From: Tom Cosgrove ***@***.***>
Date: Friday, May 3, 2024 at 10:01
To: openssl/openssl ***@***.***>
Cc: shridhar ***@***.***>, Author ***@***.***>
Subject: Re: [openssl/openssl] Move ossl_asn1_string_to_time_t() to libtestutil (PR #23269)


@t8m<https://github.com/t8m> can you reconfirm?

**Comment by t8m** [comment]
Squashed and merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**